File: src/main/java/com/koddy/server/global/config/web/WebLogConfig.java
package com.koddy.server.global.config.web;

import com.koddy.server.global.filter.MdcLoggingFilter;
import com.koddy.server.global.filter.RequestLoggingFilter;
import com.koddy.server.global.filter.RequestResponseCachingFilter;
import com.koddy.server.global.log.LoggingStatusManager;
import org.springframework.boot.web.servlet.FilterRegistrationBean;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Configuration
public class WebLogConfig {
    @Bean
    public FilterRegistrationBean<MdcLoggingFilter> firstFilter() {
        final FilterRegistrationBean<MdcLoggingFilter> registrationBean = new FilterRegistrationBean<>();
        registrationBean.setFilter(new MdcLoggingFilter());
        registrationBean.addUrlPatterns("/*");
        registrationBean.setOrder(1);
        registrationBean.setName("mdcLoggingFilter");
        return registrationBean;
    }

    @Bean
    public FilterRegistrationBean<RequestResponseCachingFilter> secondFilter() {
        final FilterRegistrationBean<RequestResponseCachingFilter> registrationBean = new FilterRegistrationBean<>();
        registrationBean.setFilter(new RequestResponseCachingFilter());
        registrationBean.addUrlPatterns("/*");
        registrationBean.setOrder(2);
        registrationBean.setName("requestResponseCachingFilter");
        return registrationBean;
    }

    @Bean
    public FilterRegistrationBean<RequestLoggingFilter> thirdFilter(final LoggingStatusManager loggingStatusManager) {
        final FilterRegistrationBean<RequestLoggingFilter> registrationBean = new FilterRegistrationBean<>();
        registrationBean.setFilter(new RequestLoggingFilter(
                loggingStatusManager,
                "/favicon.ico",
                "/error*",
                "/api/swagger*",
                "/api-docs*",
                "/actuator*"
        ));
        registrationBean.addUrlPatterns("/*");
        registrationBean.setOrder(3);
        registrationBean.setName("requestLoggingFilter");
        return registrationBean;
    }
}

File: src/main/java/com/koddy/server/global/config/web/WebLogConfig.kt
package com.koddy.server.global.config.web

import com.koddy.server.global.filter.MdcLoggingFilter
import com.koddy.server.global.filter.RequestLoggingFilter
import com.koddy.server.global.filter.RequestResponseCachingFilter
import com.koddy.server.global.log.LoggingStatusManager
import org.springframework.boot.web.servlet.FilterRegistrationBean
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

@Configuration
class WebLogConfig {
    @Bean
    fun firstFilter(): FilterRegistrationBean<MdcLoggingFilter> {
        return FilterRegistrationBean<MdcLoggingFilter>().apply {
            order = 1
            filter = MdcLoggingFilter()
            setName("mdcLoggingFilter")
            addUrlPatterns("/api/*")
        }
    }

    @Bean
    fun secondFilter(): FilterRegistrationBean<RequestResponseCachingFilter> {
        return FilterRegistrationBean<RequestResponseCachingFilter>().apply {
            order = 2
            filter = RequestResponseCachingFilter()
            setName("requestResponseCachingFilter")
            addUrlPatterns("/api/*")
        }
    }

    @Bean
    fun thirdFilter(loggingStatusManager: LoggingStatusManager): FilterRegistrationBean<RequestLoggingFilter> {
        val ignoredUrls: List<String> =
            listOf(
                "/favicon.ico",
                "/error*",
                "/api/swagger*",
                "/api-docs*",
                "/api/actuator*",
                "/api/health",
            )

        return FilterRegistrationBean<RequestLoggingFilter>().apply {
            order = 3
            filter = RequestLoggingFilter(loggingStatusManager, *ignoredUrls.toTypedArray())
            setName("requestLoggingFilter")
            addUrlPatterns("/api/*")
        }
    }
}
